lib/testing/src/equality/equality.dart

```diff
@@ -0,0 +1,229 @@
+// Copyright 2014 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+part of quiver.testing.equality;
+
+
+/**
+ * Matcher for == and hashCode methods of a class.
+ *
+ * To use, invoke areEqualityGroups with a list of equality groups where each
+ * group contains objects that are supposed to be equal to each other, and
+ * objects of different groups are expected to be unequal. For example:
+ *
+ * expect({
+ *      'hello': ["hello", "h" + "ello"],
+ *      'world': ["world", "wor" + "ld"],
+ *      'three': [2, 1 + 1]
+ *     }, areEqualityGroups);
```

Comment:
I feel `areEqualityGroups` is more precise. Ok with keeping it. In fact I'm thinking we could be even more precise, e.g. `distinctEqualityGroups` to underscore non-equality across groups.